core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/ArrayBindingSet.java

```diff
@@ -295,8 +330,8 @@ public ArrayBindingSetIterator() {
 
 		@Override
 		public boolean hasNext() {
-			for (; index < values.length; index++) {
-				if (whichBindingsHaveBeenSet[index] && values[index] != null) {
+			for (int at = index; at < values.length; at++) {
+				if (whichBindingsHaveBeenSet[at] && values[at] != null) {
 					return true;
 				}
 			}
```

Comment:
Doesn't the old solution only increments if the current position is false/null?

Reply:
No the problem is that both hasNext and next advance the index. So with a four set values,
`hasNext, hasNext, next, hasNext, hasNext, next` consumes the iterator instead of  
`hasNext, hasNext, next, hasNext, hasNext, next , hasNext, hasNext, next, hasNext, hasNext, next`.
Note the required double hasNext call required to trigger the bug.

Reply:
I see you maybe a test. I’ll try to debug it to understand what’s going on, from just reading it I can’t quite understand how it can increment a second time.

Reply:
@hmottestad you are right it didn't I miss diagnosed the issue.